DirectProgramming/DPC++FPGA/Tutorials/Features/ac_types/ac_int/src/ac_int.cpp

```diff
@@ -0,0 +1,339 @@
+  buffer<ac_int14, 1> result(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only, no_init};
+    h.single_task<ShiftLeft>([=] { res[0] = x[0] << y[0]; });
+  });
+}
+
+// Note how the shift amount is specified with a smaller ac_int than in
+// TestShiftLeft above
+void TestEfficientShiftLeft(queue &q, const ac_int14 &a, const ac_uint4 &b,
+                            ac_int14 &c) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<ac_uint4, 1> inp2(&b, 1);
+  buffer<ac_int14, 1> result(&c, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only, no_init};
+    h.single_task<EfficientShiftLeft>([=] { res[0] = x[0] << y[0]; });
+  });
+}
+
+// The method
+//      x = y.slc<M>(n)
+// is equivalent to the VHDL behavior of
+//      x := y((M+n-1) downto n);
+// Note that only static bit widths are supported
+void TestGetBitSlice(queue &q, const ac_int14 &a, ac_uint4 &b, const int lsb) {
+  buffer<ac_int14, 1> inp1(&a, 1);
+  buffer<int, 1> inp2(&lsb, 1);
+  buffer<ac_uint4, 1> result(&b, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{inp1, h, read_only};
+    accessor y{inp2, h, read_only};
+    accessor res{result, h, write_only, no_init};
+    h.single_task<GetBitSlice>([=] { res[0] = x[0].slc<4>(y[0]); });
+  });
+}
+
+// There is a set_slc(int lsb, const ac_int<W, S> &slc) which allows the user to
+// set a bit slice as shown in the example below.
+void TestSetBitSlice(queue &q, ac_int14 &a, const ac_int4 &b, const int lsb) {
+  buffer<ac_int14, 1> buff_a(&a, 1);
+  buffer<ac_int4, 1> buff_b(&b, 1);
+  buffer<int, 1> buff_lsb(&lsb, 1);
+
+  q.submit([&](handler &h) {
+    accessor x{buff_a, h, write_only, no_init};
+    accessor y{buff_b, h, read_only};
+    accessor lsb_accessor{buff_lsb, h, read_only};
+
+    h.single_task<SetBitSlice>([=] {
+      // The set_slc method does not need to have a width specified as a
+      // template argument since the width is inferred from the width of the
+      // argument x
+      x[0].set_slc(lsb_accessor[0], y[0]);
+
+      // Bits can also be individually set as follows:
+      x[0][3] = 0;
+      x[0][2] = 0;
+      x[0][1] = 0;
+      x[0][0] = 0;
+    });
+  });
+}
+
+int main() {
+#if defined(FPGA_EMULATOR)
+  ext::intel::fpga_emulator_selector selector;
+#else
+  ext::intel::fpga_selector selector;
+#endif
+  bool passed = true;
+
+  try {
+    // create the SYCL device queue
+    queue q(selector, dpc_common::exception_handler);
+
+    // Use a fixed initial seed
+    srand(123);
+
+    // Initialize two random ints
+    int t1 = rand();
+    int t2 = rand();
+    // Truncate each of the two ints to 13 bits. The 14th bit is the sign bit.
+    // In this testbench, even though the datatypes are signed, we will be
+    // storing unsigned values to make the testing process simpler.
+    t1 &= (1 << 13) - 1;
```

Comment:
Maybe make `(1 << 13) - 1` a constexpr since you use it so much?
Maybe something like:
```
constexpr int min_bits = 13;
constexpr int mask = (1 << min_bits) - 1;
```

Then `(1 << 13) - 1` everywhere is replaced with `mask`.